src/Microsoft.Identity.Web/WebApiExtensions/HasScopeHandler.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authorization;
+
+namespace Microsoft.Identity.Web
+{
+    /// <summary>
+    ///  Authorizer handler which needs to be called for the ScopeAuthorizationRequirement requirement type.
+    /// </summary>
+    internal class HasScopeHandler : AuthorizationHandler<ScopeAuthorizationRequirement>
+    {
+        /// <summary>
+        /// Makes a decision if authorization is allowed based on the scope requirements specified.
+        /// </summary>
+        /// <param name="context">The authorization context.</param>
+        /// <param name="requirement">The requirement to evaluate.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        protected override Task HandleRequirementAsync(AuthorizationHandlerContext context, ScopeAuthorizationRequirement requirement)
+        {
+            if (context.User != null && requirement.AllowedValues != null)
+            {
+                Claim? scopeClaim = context.User.FindFirst(ClaimConstants.Scp);
+
+                // Fallback to Scope claim name
+                if (scopeClaim == null)
+                {
+                    scopeClaim = context.User.FindFirst(ClaimConstants.Scope);
+                }
+
+                if (scopeClaim != null && scopeClaim.Value.Split(' ').Intersect(requirement.AllowedValues).Any())
```

Comment:
`Enumerable.Intersect` creates and mutates a [new `HashSet<>`](https://github.com/dotnet/runtime/blob/ebba1d4acb7abea5ba15e1f7f69d1d1311465d16/src/libraries/System.Linq/src/System/Linq/Intersect.cs#L76-L87) every time it is invoked. Could the allowed scopes be pre-processed to a `HashSet<>` in the `ScopeAuthorizationRequirement` ctor and then looped over here?